statbot/crawler.py

```diff
@@ -11,165 +11,201 @@
+    @abc.abstractmethod
+    async def read(self, source, last_id):
+        pass
 
-                            self.progress[channel.id] = mhist
+    @abc.abstractmethod
+    async def write(self, trans, events):
+        pass
 
-        # Remove deleted channels from tracker
-        for channel in set(self.progress.keys()) - set(self.channels.keys()):
-            del self.progress[channel.id]
+    @abc.abstractmethod
+    async def update(self, trans, source, last_id):
+        pass
 
     def start(self):
-        self.client.hooks['on_guild_channel_create'] = self._channel_create_hook
-        self.client.hooks['on_guild_channel_delete'] = self._channel_delete_hook
-        self.client.hooks['on_guild_channel_update'] = self._channel_update_hook
-
         self.client.loop.create_task(self.producer())
         self.client.loop.create_task(self.consumer())
 
     async def producer(self):
-        self.logger.info("Producer coroutine started!")
+        self.logger.info(f"{self.name}: producer coroutine started!")
 
         # Setup
         await self.client.wait_until_ready()
-        await self._init_channels()
+        await self.init()
 
         yield_delay = self.config['crawler']['yield-delay']
         long_delay = self.config['crawler']['empty-source-delay']
 
+        done = {source: False for source in self.progress}
         while True:
-            # tuple() is necessary since the underlying
-            # dict of channels may change size during
-            # an iteration
-            all_empty = True
-            for cid in tuple(self.progress.keys()):
-                # Do round-robin between all the channels
+            # Round-robin between all sources:
+            # Tuple because the underlying dictionary may change size
+            for source, last_id in tuple(self.progress.items()):
+                if done[source]:
+                    continue
+
                 try:
-                    channel = self.channels[cid]
-                    mhist = self.progress[cid]
-                    all_empty &= not await self._read(channel, mhist)
+                    events = await self.read(source, last_id)
+                    if events is None:
+                        # This source is exhausted
+                        done[source] = True
+                        await self.queue.put((source, None, NOW_ID))
+                        self.progress[source] = NOW_ID
+                    else:
+                        # This source still has more
+                        last_id = self.get_last_id(events)
+                        await self.queue.put((source, events, last_id))
+                        self.progress[source] = last_id
                 except Exception:
-                    self.logger.error(f"Error reading (or syncing) messages from channel id {cid}", exc_info=1)
+                    self.logger.error(f"Error reading events from source {source}", exc_info=1)
```

Comment:
this could be just `except`

don't you want to completely bail from this source rather than get right back in the loop?

Reply:
Well the problem might not be with the source. Perhaps one particular message has some unusual property I hadn't thought about that prevents it from being inserted, but otherwise the crawling is totally fine.

Reply:
that is possible, but we're catching a very broad exception. let me rephrase:

if a bug is preventing us from crawling a source, doesn't that ruin everything and shouldn't we fix that by bailing from this loop? alternatively, given that we don't even know why this might fail, instead of catching the exception shouldn't we just bail from the program entirely?

Reply:
Yes, I understand why we might want to have it fail on exceptions, but the possibility of otherwise innocuous SQL issues is possible.

To try and strike a balance here, I'm thinking we could have it catch and log SQLAlchemy exceptions, but fail on everything else.

Reply:
Alright I have changed this. Now only the consumer can catch exceptions, and when it does, only SQL-related things are caught. Any other kind of exception will terminate the bot.